ambry-store/src/main/java/com.github.ambry.store/BlobStoreCompactor.java

```diff
@@ -0,0 +1,885 @@
+   * the compacted segments out for the swap spaces. Returns once the compaction is complete.
+   * @param details the {@link CompactionDetails} for the compaction.
+   * @throws IllegalArgumentException if any of the provided segments doesn't exist in the log or if one or more offsets
+   * in the segments to compact are in the journal.
+   * @throws IllegalStateException if the compactor has not been initialized.
+   * @throws IOException if there were I/O errors creating the {@link CompactionLog}
+   * @throws StoreException if there were exceptions reading to writing to store components.
+   */
+  void compact(CompactionDetails details) throws IOException, StoreException {
+    if (srcIndex == null) {
+      throw new IllegalStateException("Compactor has not been initialized");
+    }
+    checkSanity(details);
+    compactionLog = new CompactionLog(dataDir.getAbsolutePath(), storeId, time, details);
+    resumeCompaction();
+  }
+
+  /**
+   * Resumes compaction from where it was left off.
+   * @throws IllegalStateException if the compactor has not been initialized or if there is no compaction to resume.
+   * @throws StoreException if there were exceptions reading to writing to store components.
+   */
+  void resumeCompaction() throws StoreException {
+    if (srcIndex == null) {
+      throw new IllegalStateException("Compactor has not been initialized");
+    } else if (compactionLog == null) {
+      throw new IllegalStateException("There is no compaction to resume");
+    }
+    runningLatch = new CountDownLatch(1);
+    /*
+    A single compaction job could be performed across multiple compaction cycles (if there aren't enough swap spaces to
+    complete the job in one cycle). Therefore, we loop and perform PREPARE, COPY, COMMIT and CLEANUP until the
+    CompactionLog reports the job as DONE.
+    Short description of each of the phases:
+    1. PREPARE - Initial state of a compaction cycle. Signifies that the cycle has just started.
+    2. COPY - Valid data from the segments under compaction is copied over to the swap spaces. If all the
+    data cannot be copied, it auto splits the compaction job into two cycles.
+    3. COMMIT - Adds the new log segments to the application log and atomically changes the set of index segments
+    4. CLEANUP - Cleans up the old log segments and their index files
+     */
+
+    try {
+      while (isActive && !compactionLog.getCompactionPhase().equals(CompactionLog.Phase.DONE)) {
+        CompactionLog.Phase phase = compactionLog.getCompactionPhase();
+        switch (phase) {
+          case PREPARE:
+            compactionLog.markCopyStart();
+            // fall through to COPY
+          case COPY:
+            copy();
+            if (isActive) {
+              compactionLog.markCommitStart();
+              commit(false);
+              compactionLog.markCleanupStart();
+              waitTillRefCountZero();
```

Comment:
I had it in cleanup before but waiting for ref count to be zero isn't a cleanup task. This wait is required in the normal crashless scenario of compaction. During recovery, the wait isn't required. The two options were to do this or have a 
```
if (!recovering) {
  waitTillRefCountZero();
}
```
block in cleanup. I went for this because I find this more intuitive. I don't think having one method for each of the tasks is a requirement,